src/Logic.hs

```diff
@@ -91,150 +88,122 @@ import qualified Time
-  | IsReviewer Username (Bool -> a)
-  | GetPullRequest PullRequestId (Maybe GithubApi.PullRequest -> a)
-  | GetOpenPullRequests (Maybe IntSet -> a)
-  | GetLatestVersion Sha (Either TagName Integer -> a)
-  | GetChangelog TagName Sha (Maybe Text -> a)
-  | IncreaseMergeMetric a
-  | UpdateTrainSizeMetric Int a
-  deriving (Functor)
-
-data RetrieveEnvironmentFree a
-  = GetProjectConfig (ProjectConfiguration -> a)
-  | GetDateTime (UTCTime -> a)
-  | GetBaseBranch (BaseBranch -> a)
-  deriving (Functor)
-
-type BaseAction = Free BaseActionFree
-type RetrieveEnvironment = Free RetrieveEnvironmentFree
-
--- | The actions represented here will be tranformed into a lower-level type of actions,
--- embodied in the Operation monad.
-type Action = Free (Sum BaseActionFree RetrieveEnvironmentFree)
-
-type Operation = Free (Sum
-                        (Sum
-                          MetricsOperationFree
-                          TimeOperationFree)
-                        (Sum
-                          GitOperationFree
-                          GithubOperationFree))
+    } -> BaseAction m (Either IntegrationFailure Sha)
+  TryPromote :: Branch -> Sha -> BaseAction m PushResult
+  TryPromoteWithTag :: Branch -> Sha -> TagName -> TagMessage -> BaseAction m PushWithTagResult
+  CleanupTestBranch :: PullRequestId -> BaseAction m ()
+  LeaveComment :: PullRequestId -> Text -> BaseAction m ()
+  IsReviewer :: Username -> BaseAction m Bool
+  GetPullRequest :: PullRequestId -> BaseAction m (Maybe GithubApi.PullRequest)
+  GetOpenPullRequests :: BaseAction m (Maybe IntSet)
+  GetLatestVersion :: Sha -> BaseAction m (Either TagName Integer)
+  GetChangelog :: TagName -> Sha -> BaseAction m (Maybe Text)
+  IncreaseMergeMetric :: BaseAction m ()
+  UpdateTrainSizeMetric :: Int -> BaseAction m ()
+
+type instance DispatchOf BaseAction = 'Dynamic
+
+data RetrieveEnvironment :: Effect where
+  GetProjectConfig :: RetrieveEnvironment m ProjectConfiguration
+  GetDateTime :: RetrieveEnvironment m UTCTime
+  GetBaseBranch :: RetrieveEnvironment m BaseBranch
+
+type instance DispatchOf RetrieveEnvironment = 'Dynamic
 
 data PRCloseCause =
       User            -- ^ The user closed the PR.
     | StopIntegration -- ^ We close and reopen the PR internally to stop its integration if it is approved.
   deriving Show
 
-doBaseAction :: BaseAction a -> Action a
-doBaseAction = hoistFree InL
-
-doEnvironment :: RetrieveEnvironment a -> Action a
-doEnvironment = hoistFree InR
-
-liftAction :: BaseActionFree a -> Action a
-liftAction = doBaseAction . liftF
-
-liftEnvironment :: RetrieveEnvironmentFree a -> Action a
-liftEnvironment = doEnvironment . liftF
-
-doTime :: TimeOperation a -> Operation a
-doTime = hoistFree (InL . InR)
-
-doGit :: GitOperation a -> Operation a
-doGit = hoistFree (InR . InL)
-
-doGithub :: GithubOperation a -> Operation a
-doGithub = hoistFree (InR . InR)
-
-doMetrics :: MetricsOperation a -> Operation a
-doMetrics = hoistFree (InL . InL)
```

Comment:
It was satisfying to be able to remove the above code. With effectful, we're not manually managing trees of effects, so no lifting, hoisting and other synonyms for raising things.